src/freenas/usr/local/lib/middlewared_truenas/plugins/failover_/event_linux.py

```diff
@@ -0,0 +1,713 @@
+                    ('method', '=', 'failover.events.vrrp_backup')
+                ])
+            ]
+        )
+
+        # only care about RUNNING events
+        current_events = [i for i in current_events if i['state'] == 'RUNNING']
+        for i in current_events:
+            if i['method'] == 'failover.events.vrrp_master':
+                # if the incoming event is also a MASTER event then log it and ignore
+                if event in ('MASTER', 'forcetakeover'):
+                    logger.warning(
+                        'A failover MASTER event is already being processed, ignoring.'
+                    )
+                    raise IgnoreFailoverEvent()
+
+            if i['method'] == 'failover.events.vrrp_backup':
+                # if the incoming event is also a BACKUP event then log it and ignore
+                if event == 'BACKUP':
+                    logger.warning(
+                        'A failover BACKUP event is already being processed, ignoring.'
+                    )
+                    raise IgnoreFailoverEvent()
+
+            # TODO: timdelta flapping event
+
+    def _event(self, ifname, event):
+
+        forcetakeover = False
+        if event == 'forcetakeover':
+            forcetakeover = True
+
+        # generate data to be used during the failover event
+        fobj = self.generate_failover_data()
+
+        if not forcetakeover:
+            if fobj['disabled'] and not fobj['master']:
+                # if forcetakeover is false, and failover is disabled
+                # and we're not set as the master controller, then
+                # there is nothing we need to do.
+                logger.warning(
+                    'Failover is disabled but this node is marked as the BACKUP node. Assuming BACKUP.'
+                )
+                raise IgnoreFailoverEvent()
+
+            # If there is a state change on a non-critical interface then
+            # ignore the event and return
+            ignore = [i for i in fobj['non_crit_interfaces'] if i in ifname]
+            if ignore:
+                logger.warning(
+                    'Ignoring state change on non-critical interface "%s".', ifname
+                )
+                raise IgnoreFailoverEvent()
+
+            # this section needs to run as quick as possible so we check if the remote
+            # client is even connected before we try and start doing remote_calls
+            remote_connected = self.run_call('failover.remote_connected')
+
+            # if the other controller is already master, then assume backup
+            if remote_connected:
+                if self.run_call('failover.call_remote', 'failover.status') == 'MASTER':
+                    logger.warning('Other node is already MASTER, assuming BACKUP.')
+                    raise IgnoreFailoverEvent()
+
+            # ensure the zpools are imported
+            needs_imported = False
+            for vol in fobj['volumes']:
```

Comment:
Lets do a single pool.query call?

for pool in self.run_call('pool.query', [('name', 'in', fobj['volumes'])])